tests/sampler/test_parscan.cpp

```diff
@@ -0,0 +1,83 @@
+#include <iostream>
+#include <sstream>
+#include <vector>
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+#include "Kokkos_Core.hpp"
+
+using ::testing::HasSubstr;
+using ::testing::Not;
+
+struct Tester {
+  template <typename execution_space>
+  explicit Tester(const execution_space& space) {
+    //! Explicitly launch a kernel with a name, and run it 150 times with kernel
+    //! logger. Use a periodic sampling with skip rate 51. This should print
+    //! out 2 invocations, and there is a single matcher with a regular
+    //! expression to check this.
+
+    long int N = 1024;
+    int64_t result;
+
+    for (int iter = 0; iter < 150; iter++) {
+      result = 0;
+      Kokkos::parallel_scan("named kernel scan", N, *this, result);
+    }
+  }
+
+  KOKKOS_FUNCTION void operator()(const int, long int&, bool) const {}
+};
+
+static const std::vector<std::string> matchers{
+    "KokkosP: sample 51 calling child-begin function...",
+    "KokkosP: sample 51 finished with child-begin function.",
+    "KokkosP: sample 51 calling child-end function...",
+    "KokkosP: sample 51 finished with child-end function.",
+    "KokkosP: sample 102 calling child-begin function...",
+    "KokkosP: sample 102 finished with child-begin function.",
+    "KokkosP: sample 102 calling child-end function...",
+    "KokkosP: sample 102 finished with child-end function."};
+
+/**
+ * @test This test checks that the tool effectively samples.
+ *
+
+ */
+
+TEST(SamplerTest, ktoEnvVarDefault) {
+  //! Initialize @c Kokkos.
+  Kokkos::initialize();
+
+  //! Redirect output for later analysis.
+  std::cout.flush();
+  std::ostringstream output;
+  std::streambuf* coutbuf = std::cout.rdbuf(output.rdbuf());
+
+  //! Run tests. @todo Replace this with Google Test.
+  Tester tester(Kokkos::DefaultExecutionSpace{});
+
+  //! Finalize @c Kokkos.
+  Kokkos::finalize();
+
+  //! Restore output buffer.
+  std::cout.flush();
+  std::cout.rdbuf(coutbuf);
+  std::cout << output.str() << std::endl;
+
+  //! Analyze test output.
+  for (const auto& matcher : matchers) {
+    EXPECT_THAT(output.str(), HasSubstr(matcher));
+  }
+
+  EXPECT_THAT(output.str(), Not(HasSubstr("KokkosP: sample 1 calling")));
+  EXPECT_THAT(output.str(), Not(HasSubstr("KokkosP: sample 75 calling")));
+  EXPECT_THAT(output.str(), Not(HasSubstr("KokkosP: sample 149 calling")));
```

Comment:
I would rather do
```suggestion
   int occurrences = 0;
   while (std::string::size_type pos = 0; (pos = s.find("calling child-begin function", pos )) != std::string::npos) {
          ++occurrences;
          pos += target.length();
   }
   EXPECT_EQ(occurrences, 3);
```

Reply:
Nice. But I think this should be two occurrences and not three, right? Specifically, I think you intended the parameter 3 in the function call in the last line to be 2. Only samples 51 and 102 are taken. 

And, why not have both your and my test in the code rather than replacing with yours? I think my code that you deleted is still covering different case (that a particular sample number is _not_ taken), so our two tests are complementary, right?

Reply:
> Nice. But I think this should be two occurrences and not three, right? Specifically, I think you intended the parameter 3 in the function call in the last line to be 2. Only samples 51 and 102 are taken.

Sure.

> And, why not have both your and my test in the code rather than replacing with yours? I think my code that you deleted is still covering different case (that a particular sample number is _not_ taken), so our two tests are complementary, right?

I don't think we need both with what the tools are currently printing but it's fine with me if you want to have both.

Reply:
OK, got it. 

I will commit both (with the change of 2 occurrences for your suggested change) for now.